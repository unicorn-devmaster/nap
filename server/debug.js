// Type
const LOG = '#'
const INFO = '$'
const DEBUG = '@'
const WARN = '!'
const ERROR = '*'

class debug extends console.Console { // eslint-disable-line
  constructor() {
    super();
  }

  static _toString(any) {
    // Can be crash somehow eg. circular object JSON
    let result = ''
    any.forEach(item => {
      try {
        result += (item.constructor === String) ? item : JSON.stringify(item, null, 2)
      } catch (err) {
        // NVM
      }
    })

    return result
  }

  static _print(type, text) {
    // No console in prod
    if (process.env.NODE_ENV === 'production') return global.debug

    try {
      const at = new Date()

      console.log([ // eslint-disable-line
        at.toISOString(),
        type,
        text
      ].join(' | '))
    } catch (err) {
      console.error(err) // eslint-disable-line
    }

    return global.debug
  }

  static log(...any) {
    const text = debug._toString(any)
    return debug._print(LOG, text)
  }

  static info(...any) {
    return debug._print(INFO, any)
  }

  static warn(...any) {
    return debug._print(WARN, any)
  }

  static error(...any) {
    return debug._print(ERROR, any)
  }

  static debug(...any) {
    return debug._print(DEBUG, any)
  }
}

global.debug = debug