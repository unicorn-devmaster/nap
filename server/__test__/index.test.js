/* eslint-env jest */
require('../debug')

describe('index', () => {

  const fetcher = async (body, authorization) => {
    const fetch = require('isomorphic-fetch')
    return await fetch('http://localhost:3000/graphql', {
      method: 'post',
      headers: { 'Content-Type': 'application/json', authorization },
      body
    }).then(response => response.json())
  }

  it('has GraphQL', async () => {
    await fetcher().then(res => expect(res).toMatchSnapshot())
  })

  const loginWithFacebook = {
    operationName: 'loginWithFacebook',
    query: `mutation loginWithFacebook($deviceInfo: String!, $accessToken: String!) {
      loginWithFacebook(deviceInfo: $deviceInfo, accessToken: $accessToken) {
        isLoggedIn
        sessionToken
        user {
          _id
          name
          status
        }
      }
      errors {
        code
        message
      }
    }`,
    variables: `{"deviceInfo": "foo", "accessToken": "EAABnTrZBSJyYBAKvcWAcAOUwt07ZCVxhCYQwKKWFZAwtOhsGYZAc7olL04W8eJTlxBeZCmxCQO9kYZA4kKtTD0zmZChhb5hEoZBl7JHT0Rx39uGP8ow2X9vGoTLFZCm4Dd0NFvH0qsHXNYinsOKjszfSJVOj3DZChv0MNszawr1le8O0ToqI3Ak9Jr8X3X6imEtvJ2q8ceeVh5Ux1rSbgypRQNRDjlredVXpIZD"}`
  }

  it('can log user in with Facebook token', async () => {
    const query = JSON.stringify(loginWithFacebook)

    await fetcher(query).then(result => {
      const sessionToken = result.data.loginWithFacebook.sessionToken
      expect(result).toMatchObject({
        "data": {
          "loginWithFacebook": {
            sessionToken,
            "user": {
              "_id": expect.any(String),
              "name": expect.any(String),
            }
          },
          "errors": null
        }
      })
    })
  })

  it('can log user out and return null user', async () => {
    // Login first
    const loginWithFacebook_query = JSON.stringify(loginWithFacebook)
    const sessionToken = await fetcher(loginWithFacebook_query).then(result => result.data.loginWithFacebook.sessionToken)

    // Then logout
    const logout = {
      query: `
mutation {
  logout {
    loggedOutAt
    isLoggedIn
  }
  errors {
    code
    message
  }
}`,
      variables: null
    }

    const query = JSON.stringify(logout)
    const authorization = `Bearer ${sessionToken}`

    const result = await fetcher(query, authorization)
    expect(result).toMatchObject({
      "data": {
        "logout": {
          "loggedOutAt": expect.any(String),
          "isLoggedIn": false,
        },
        "errors": null
      }
    })

    const userProfile = {
      query: `query {
        authen {
          isLoggedIn
        }
        user {
          _id
          name
          status
        }
        errors {
          code
          message
        }
      }`,
      variables: null
    }

    const userProfile_query = JSON.stringify(userProfile)
    const userProfile_result = await fetcher(userProfile_query)
    expect(userProfile_result).toMatchObject({
      "data": {
        "authen": {
          "isLoggedIn": false,
        },
        "errors": [
          {
            "code": 0,
            "message": "No session found"
          }
        ],
        "user": null
      }
    })
  })
})