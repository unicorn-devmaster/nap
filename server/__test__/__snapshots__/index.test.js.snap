// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`index should have start function 1`] = `[Function]`;
